Sources/MapboxNavigation/OverlayController.swift

```diff
@@ -0,0 +1,390 @@
+        }
+        
+        if let compositeShieldImage = currentShieldName, let roadName = currentRoadName {
+            let compositeShield = NSMutableAttributedString(string: " \(roadName)")
+            compositeShield.insert(compositeShieldImage, at: 0)
+            currentShieldName = compositeShield
+        }
+        
+        return (roadName: currentRoadName, shieldName: currentShieldName)
+    }
+    
+    /**
+     Method, which returns list of source identifiers, which contain streets tile set.
+     */
+    private func streetsSources() -> [StyleObjectInfo] {
+        let streetsSources = (try? navigationMapView.mapView.__map.getStyleSources().compactMap {
+            $0
+        }.filter {
+            let identifiers = tileSetIdentifiers($0.id, sourceType: $0.type)
+            return isMapboxStreets(identifiers)
+        }) ?? []
+        
+        return streetsSources
+    }
+    
+    /**
+     Method, which returns identifiers of the tile sets that make up specific source.
+     
+     This array contains multiple entries for a composited source. This property is empty for non-Mapbox-hosted tile sets and sources with type other than `vector`.
+     */
+    private func tileSetIdentifiers(_ sourceIdentifier: String, sourceType: String) -> [String] {
+        do {
+            if sourceType == "vector",
+               let properties = try navigationMapView.mapView.__map.getStyleSourceProperties(forSourceId: sourceIdentifier).value as? Dictionary<String, Any>,
+               let url = properties["url"] as? String,
+               let configurationURL = URL(string: url),
+               configurationURL.scheme == "mapbox",
+               let tileSetIdentifiers = configurationURL.host?.components(separatedBy: ",") {
+                return tileSetIdentifiers
+            }
+        } catch {
+            NSLog("Failed to get source properties with error: \(error.localizedDescription).")
+        }
+        
+        return []
+    }
+    
+    private func roadShieldTextColor(line: Feature) -> UIColor? {
+        guard let shield = line.properties?["shield"] as? String else {
+            return nil
+        }
+        
+        // shield_text_color is present in Mapbox Streets source v8 but not v7.
+        guard let shieldTextColor = line.properties?["shield_text_color"] as? String else {
+            let currentShield = HighwayShield.RoadType(rawValue: shield)
+            return currentShield?.textColor
+        }
+        
+        switch shieldTextColor {
+        case "black":
+            return .black
+        case "blue":
+            return .blue
+        case "white":
+            return .white
+        case "yellow":
+            return .yellow
+        case "orange":
+            return .orange
+        default:
+            return .black
+        }
+    }
+    
+    private func roadShieldAttributedText(for text: String, textColor: UIColor, imageName: String) -> NSAttributedString? {
```

Comment:
This method should live on NavigationMapView.

Reply:
Having this as part of `NavigationMapView` is not straightforward. This method is used only as a part `WayNameView` setup, and since it's setup is quite specific, having it in `NavigationMapView` seems incorrect. It is a part of a custom setup method, which may not be needed anywhere else, and also doing such would make `WayNameView` dependent on `NavigationMapView`. Instead I've decided to move it to `WayNameView` extension together with other setup methods